File: examples/simple.c
/* simple.c -- basic ppamltracer example
 * This file is in the public domain.
 *
 * Compile this with
 *     cc -std=c99 -lppamltracer -o simple simple.c
 */

#include <stdio.h>
#include <stddef.h>

#include <ppaml/tracer.h>

// ppamltracer state
ppaml_tracer_t tracer;
ppaml_phase_t phase_fact;
ppaml_phase_t phase_fib;


// Prototypes
unsigned long long fact(const unsigned short);
unsigned long fib(const unsigned short);


int main(int argc, char *argv[])
{
	/* Disable buffering on stdout so we can see the numbers as they are
	 * printed. */
	setvbuf(stdout, NULL, _IONBF, 0);

	// Start ppamltracer.
	ppaml_tracer_init(&tracer, "/tmp/simple_report");

	// Register the phases.
	ppaml_phase_init(&tracer, &phase_fact, "fact");
	ppaml_phase_init(&tracer, &phase_fib, "fib");

	// Print factorials.
	printf("Factorials: ");
	for (unsigned short i = 0; i < 21; i++) {
		printf("%lld ", fact(i));
	}
	printf("\n");

	// Print Fibonacci numbers.
	printf("Fibonacci numbers: ");
	for (unsigned short i = 0; i < 28; i++) {
		printf("%ld ", fib(i));
	}
	printf("\n");

	// Finalize ppamltracer and exit.
	ppaml_phase_done(&phase_fact);
	ppaml_phase_done(&phase_fib);
	ppaml_tracer_done(&tracer);
	return 0;
}

unsigned long long fact(const unsigned short n)
{
	// Record that we have entered the 'fact' phase.
	ppaml_phase_start(&phase_fact);

	// Compute the factorial.
	unsigned long long result;
	if (n == 0) {
		result = 1;
	} else {
		result = (unsigned long long)n * fact(n - 1);
	}

	// Record that we're exiting the 'fact' phase.
	ppaml_phase_stop(&phase_fact);
	return result;
}

unsigned long fib(const unsigned short n)
{
	// Record that we have entered the 'fib' phase.
	ppaml_phase_start(&phase_fib);

	// Compute the nth Fibonancci number.
	unsigned long result;
	if (n == 0 || n == 1) {
		result = 1;
	} else {
		result = fib(n - 1) + fib(n - 2);
	}

	// Record that we're exiting the 'fact' phase.
	ppaml_phase_stop(&phase_fib);
	return result;
}

File: bindings/java/examples/src/com/galois/ppaml/tracer/examples/Simple.java
/* Simple -- basic ppamltracer example
 * This file is in the public domain. */

package com.galois.ppaml.tracer.examples;

import com.galois.ppaml.tracer.ExplicitlyConfiguredTracer;
import com.galois.ppaml.tracer.Phase;
import com.galois.ppaml.tracer.Tracer;

public class Simple {

	protected Phase factPhase;

	protected Phase fibPhase;

	protected void go(Tracer tracer) {
		// Register the phases.
		factPhase = tracer.createPhase("fact");
		fibPhase = tracer.createPhase("fib");

		// Print factorials.
		System.out.print("Factorials:");
		for (int i = 0; i < 21; i++) {
			System.out.print(" " + fact(i));
		}
		System.out.println();

		// Print Fibonacci numbers.
		System.out.print("Fibonacci numbers:");
		for (int i = 0; i < 30; i++) {
			System.out.print(" " + fib(i));
		}
		System.out.println();

		// Finalize phases and exit.
		factPhase.close();
		fibPhase.close();
	}

	protected long fact(int n) {
		// Record that we're running inside the factorial phase.
		factPhase.start();
		// Compute the factorial.
		long result;
		if (n == 0) {
			result = 1;
		} else {
			result = n * fact(n - 1);
		}
		// Return the result.
		factPhase.stop();
		return result;
	}

	protected int fib(int n) {
		// Record that we're running inside the Fibonacci phase.
		fibPhase.start();
		// Compute the nth Fibonacci number.
		int result;
		switch (n) {
		case 0:
		case 1:
			result = n;
			break;
		default:
			result = fib(n - 1) + fib(n - 2);
			break;
		}
		// Return the result.
		fibPhase.stop();
		return result;
	}

	public static void main(String[] args) {
		// Start ppamltracer.
		Tracer tracer = new ExplicitlyConfiguredTracer("/tmp/simple_report");

		// Run the app.
		(new Simple()).go(tracer);

		// Finalize ppamltracer and exit.
		tracer.close();
	}

}
